Experimental: support for uv

antirotor opened the pull request:
## Changelog Description
[uv](https://astral.sh/blog/uv) is an extremely fast Python package installer and resolver, written in Rust, and designed as a drop-in replacement for pip and pip-tools workflows. This PR is trying experimentaly use uv instead of Poetry to handle dependencies.

## Additional info
Thanks to it's speed, we could use it to dynamically build environments for every process we run (ie. when running DCC with different Python, to avoid python dependency conflicts with main process).

> [!NOTE]
This is just an experiment - to see how uv can be utilized and if it provides something more than Poetry already does. It has it's own caveats - lock files are platform specific for example, and so on.

### Testing:
running manage.ps1 should utilize uv already - **note** linux version isn't done yet.

tweak-wtf commented:
i just checked this out and fiddled around a bit to get it running on my windows box. naively running `./tools/manage.ps1 create-env` until it was working. had uv installed already the same way as in `manage.ps1`. 

here's the changes: https://github.com/ynput/ayon-launcher/compare/feature/use-uv...tweak-wtf:ayon-launcher:feature/use-uv-playground

it was picking up my system python so i added a python flag in `uv venv` that would restrict to 3.9.
then it was failing for not finding `.venv/Scripts/pre-commit` as it somehow didn't got installed from the compiled `requirements.txt`. so adding it to `windows-requirements.txt` did the trick.

i do get version warnings when running the tray because i think it's checking my system python but seems cosmetic. anyways, with cached dependencies the install takes 3s for me rn😅

antirotor commented:
> EDIT: i realized the `.python-version` file only in the end. i think this would be better than adding the python flag to `uv venv`

Ah, I've never get into that issue because I am always running **pyenv** so there's always `.python-version`. Maybe the next step would be to change https://github.com/ynput/ayon-dependencies-tool to make use of uv. I am wondering what speed gains could we find there.

antirotor commented:
So there is another use case perfect for uv - we can solve some of the issues like [this one](https://github.com/ynput/ayon-unreal/issues/158) or in general all cases where we now install pyside2/6 and otio to running python environment.  To support offline studios, we need to distribute "cached" pip index with AYON somehow and build venv using uv on the fly. Problem with uv and it's caching is (if I understand it correctly) that uv is not caching wheels themselves, but unzipped content and there is no "populate cache" functionality currently.

MustafaJafar commented:
Are we checking against the wrong python version?
```shell
>>> Detecting host Python ... OK [ 3.9 ]
>>> Testing venv ... OK
warning: The `requires-python` specifier (`~=3.9`) in `ayon` uses the tilde specifier (`~=`) without a patch version. This will be interpreted as `>=3.9, <4`. Did you mean `~=3.9.0` to constrain the version as `>=3.9.0, <3.10`? We recommend only using the tilde specifier with a patch version to avoid ambiguity.
error: Distribution `pywin32==310 @ registry+https://pypi.org/simple` can't be installed because it doesn't have a source distribution or wheel for the current platform

hint: You're using CPython 3.14 (`cp314`), but `pywin32` (v310) only has wheels with the following Python ABI tags: `cp39`, `cp310`, `cp311`, `cp312`, `cp313`
```

MustafaJafar commented on pyproject.toml:
```diff
@@ -1,55 +1,81 @@
+    "coolname",
+    "keyring ~= 22.0.1",
+    # environment variables resolving
+    "acre @ git+https://github.com/pypeclub/acre.git",
```
```suggestion
    "acre @ git+https://github.com/ynput/acre.git",
```

iLLiCiTiT replied:
```suggestion
```

MustafaJafar commented:
while we in this PR could we replace this line
https://github.com/ynput/ayon-launcher/blob/1af75af4a64e58d8d2b6ac5a237c8f8c023886c1/tools/ayon_console.bat#L11

with line
```
uv run start.py %*
```

iLLiCiTiT commented:
So, I would remove `acre` and `appdirs` from ayon launcher dependencies. In middle of testing but we should finally remove them.

It is strange that the requirements at the end do contain `pip` because the `pip` is not in the build (but there is pip egg, maybe that is the reason?).

iLLiCiTiT commented on pyproject.toml:
```diff
@@ -1,55 +1,81 @@
+    "Unidecode ~= 1.3.0",
+    "aiohttp ~= 3.7",
+    # local settings
+    "appdirs @ git+https://github.com/ActiveState/appdirs.git",
```
```suggestion
```

iLLiCiTiT commented on pyproject.toml:
```diff
@@ -1,55 +1,81 @@
+    "pywin32 == 310; sys_platform == 'win32'",
+    "python3-xlib; sys_platform == 'linux'",
+    "distro == 1.9.0; sys_platform == 'linux'",
+    "pip ~= 24.0",
```
```suggestion
```

iLLiCiTiT commented:
It uses python 3.9.22 on my machine, which causes that dependencies tool fails...

iLLiCiTiT commented:
> Should I be able to just run through uv - and if we don't support it. Why not?

Because we have to side install Qt to separate directory (and different version based on the needs).

BigRoy commented:
> Because we have to side install Qt to separate directory.

It doesn't answer the full question - why do we have to? And why couldn't UV (or whatever startup procedure we have) do it for us?

We want to repo entry point, like `uv run` make it as simple as possible to get up and running.
This PR also does not describe the actual commands I need to run to get a fully working setting.

BigRoy reviewed (commented):
Overall this seems to work nicely - I'm unable to get the `build` to fully work however.

I currently just need to install the dependencies by first activating the venv so it uses the `uv` python.
```
uv sync
.venv\Scripts\activate
powershell .\tools\manage.ps1 create-env
powershell .\tools\manage.ps1 install-runtime-dependencies
```
Then I can `uv run start.py` etc.
However the `build` still fails:
```
warning: `VIRTUAL_ENV=F:\dev\ayon-launcher\.venv` does not match the project environment path `.venv` and will be ignored; use `--active` to target the active environment instead
error: Querying Python at `F:\dev\ayon-launcher\shim\.venv\Scripts\python.exe` failed with exit status exit code: 103
System.Management.Automation.RemoteException
[stderr]
No Python at 'C:\Users\User\.pyenv\pyenv-win\versions\3.9.13\python.exe'
```
I'm not sure why it still really wants to go through `pyenv`?

antirotor commented:
Also - with the pyenv/uv python version - uv suppors `.python_version` that is also used by pyenv, but also respects what is defined in `pyproject.toml` - there is that resolution order documented somewhere. IMO we should drop the check for python version from the shell scripts and let uv handle that - it will fail if you try to use incompatible version of python with it - and that would also leave support for pyenv.

iLLiCiTiT commented:
> Also - with the pyenv/uv python version - uv suppors .python_version that is also used by pyenv, but also respects what is defined in pyproject.toml - there is that resolution order documented somewhere. IMO we should drop the check for python version from the shell scripts and let uv handle that - it will fail if you try to use incompatible version of python with it - and that would also leave support for pyenv.

Removed all python detections, and locked python version in pyproject.toml to 3.9.11.

iLLiCiTiT commented:
The important bit is that if you build a version on your machine today and I build the same version on my machine tomorrow, we should get the exact same output no matter what.

MustafaJafar commented:
I pulled latest commits, things appear to work as usual.
One thing though,  `uv.lock` file content changes on my side after updating the env.

BigRoy commented on tools/build_post_process.py:
```diff
@@ -407,10 +426,9 @@ def cleanup_files(deps_dir, libs_dir):
-
+    to_delete.extend(
+        (libs_dir / "ayon", libs_dir / "ayon.pth", deps_dir / "ayon.pth")
+    )
```
Admittedly, the old way made it much easier to see the subtle differences between each entry.

antirotor commented on tools/_venv_deps.py:
```diff
@@ -8,50 +8,34 @@
+                package_version = f"git+{url}"
             else:
-                raise ValueError(f"Unknown source type {source_type}")
+                # raise ValueError(f"Unknown source type {source}")
```
To mitigate values like
```
- Unknown source type {'registry': 'https://pypi.org/simple'}
- Unknown source type {'editable': '.'}
...
```
Could be handled better. It was quick fix.

antirotor commented:
> ah damn it. no luck on mac. Build works, but can't be executed :(

it works, it was just missing signature. So I guess it is ready for testing on all platforms.

BigRoy reviewed (commented):
This will also initialize the venv with uv - taking a tiny bit of time, even though all I wanted to check is the arguments.
Can we avoid having it force uv initialization in that case?
```cmd
powershell .\tools\manage.ps1
```

---

The `create-env` argument to `manage.ps1` still mentions Poetry:

> Install Poetry and update venv by lock file

It shouldn't, right?

---

`powershell .\tools\manage.ps1 run` worked fine.

The `build` failed for me on Windows, see:

```
powershell .\tools\manage.ps1 build

                    â–„â–ˆâ–ˆâ–„
         â–„â–ˆâ–ˆâ–ˆâ–„ â–€â–ˆâ–ˆâ–„ â–€â–ˆâ–ˆâ–€ â–„â–ˆâ–ˆâ–€ â–„â–ˆâ–ˆâ–€â–€â–€â–ˆâ–ˆâ–„    â–€â–ˆâ–ˆâ–ˆâ–„      â–ˆâ–„
        â–„â–„ â–€â–ˆâ–ˆâ–„  â–€â–ˆâ–ˆâ–„  â–„â–ˆâ–ˆâ–€ â–ˆâ–ˆâ–€      â–€â–ˆâ–ˆâ–„  â–„  â–€â–ˆâ–ˆâ–„    â–ˆâ–ˆâ–ˆ
       â–„â–ˆâ–ˆâ–€  â–ˆâ–ˆâ–„   â–€ â–„â–„ â–€  â–ˆâ–ˆ         â–„â–ˆâ–ˆ  â–ˆâ–ˆâ–ˆ  â–€â–ˆâ–ˆâ–„  â–ˆâ–ˆâ–ˆ
      â–„â–ˆâ–ˆâ–€    â–€â–ˆâ–ˆâ–„   â–ˆâ–ˆ    â–€â–ˆâ–ˆâ–„      â–„â–ˆâ–ˆâ–€  â–ˆâ–ˆâ–ˆ    â–€â–ˆâ–ˆ â–€â–ˆâ–€
     â–„â–ˆâ–ˆâ–€      â–€â–ˆâ–ˆâ–„  â–€â–ˆ      â–€â–ˆâ–ˆâ–„â–„â–„â–„â–ˆâ–ˆâ–€    â–ˆâ–€      â–€â–ˆâ–ˆâ–„

     Â·  Â· - =[ by YNPUT ]:[ http://ayon.ynput.io ]= - Â·  Â·

>>> Detecting host Python ... Resolved 119 packages in 3ms
Audited 49 packages in 1ms
OK [ 3.9 ]


    Directory: F:\dev\ayon-launcher


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----        19-12-2025     23:39                build


    Directory: F:\dev\ayon-launcher\shim


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----        19-12-2025     23:39                dist
--- Cleaning build directory ...
>>> Making sure submodules are up-to-date ...
>>> AYON [ 1.4.4-dev ]
>>> Testing venv presence ... NOT FOUND
*** We need to create virtual env first ...
>>> Test if UV is installed ... OK
>>> Creating and activating venv ...
Using CPython 3.9.23
Creating virtual environment at: .venv
Activate with: .venv\Scripts\activate
Resolved 119 packages in 3ms
░░░░░░░░░░░░░░░░░░░░ [0/62] Installing wheels...                                                                                                                                                                 warning: Failed to hardlink files; falling back to full copy. This may lead to degraded performance.
         If the cache and target directories are on different filesystems, hardlinking may not be supported.
         If this is intentional, set `export UV_LINK_MODE=copy` or use `--link-mode=copy` to suppress this warning.
Installed 62 packages in 19.66s
 + alabaster==0.7.16
 + astroid==2.15.8
 + atomicwrites==1.4.1
 + autopep8==2.2.0
 + babel==2.16.0
 + cabarchive==0.2.4
 + colorama==0.4.6
 + coverage==7.6.1
 + cx-freeze==8.4.1
 + cx-logging==3.2.1
 + dill==0.3.8
 + docutils==0.19
 + flake8==6.1.0
 + gitdb==4.0.11
 + gitpython==3.1.43
 + hatchling==1.27.0
 + imagesize==1.4.1
 + importlib-metadata==8.7.0
 + iniconfig==2.0.0
 + isort==5.13.2
 + jedi==0.19.1
 + jinja2==3.1.4
 + lazy-object-proxy==1.10.0
 + lief==0.15.1
 + linkify-it-py==2.0.3
 + markdown-it-py==2.2.0
 + markupsafe==2.0.1
 + mccabe==0.7.0
 + mdit-py-plugins==0.3.5
 + mdurl==0.1.2
 + myst-parser==0.18.1
 + parso==0.8.4
 + pathspec==0.12.1
 + pluggy==1.5.0
 + py==1.11.0
 + pycodestyle==2.11.1
 + pydocstyle==6.3.0
 + pyflakes==3.1.0
 + pygments==2.18.0
 + pylint==2.17.7
 + pytest==6.2.5
 + pytest-cov==5.0.0
 + pytest-print==0.3.1
 + setuptools==65.7.0
 + smmap==5.0.1
 + snowballstemmer==2.2.0
 + sphinx==5.3.0
 + sphinxcontrib-applehelp==2.0.0
 + sphinxcontrib-devhelp==2.0.0
 + sphinxcontrib-htmlhelp==2.1.0
 + sphinxcontrib-jsmath==1.0.1
 + sphinxcontrib-qthelp==2.0.0
 + sphinxcontrib-serializinghtml==2.0.0
 + striprtf==0.0.29
 + tomli==2.0.1
 + tomlkit==0.13.2
 + trove-classifiers==2025.5.9.12
 + typing-extensions==4.12.2
 + uc-micro-py==1.0.3
 + wheel==0.44.0
 + wrapt==1.16.0
 + zipp==3.23.0
>>> Installing pre-commit hooks ...
pre-commit installed at .git\hooks\pre-commit
>>> Virtual environment created.
>>> Cleaning cache files ... OK
>>> Building AYON shim ...
------------------------------------------
warning: `VIRTUAL_ENV=F:\dev\ayon-launcher\.venv` does not match the project environment path `.venv` and will be ignored; use `--active` to target the active environment instead
Using CPython 3.13.3
Creating virtual environment at: .venv
warning: Failed to hardlink files; falling back to full copy. This may lead to degraded performance.
         If the cache and target directories are on different filesystems, hardlinking may not be supported.
         If this is intentional, set `export UV_LINK_MODE=copy` or use `--link-mode=copy` to suppress this warning.
Installed 11 packages in 3.62s
Traceback (most recent call last):
  File "F:\dev\ayon-launcher\shim\setup.py", line 69, in <module>
    Executable(
    ~~~~~~~~~~^
        "shim_start.py",
        ^^^^^^^^^^^^^^^^
    ...<2 lines>...
        icon=icon_path
        ^^^^^^^^^^^^^^
    ),
    ^
  File "F:\dev\ayon-launcher\shim\.venv\Lib\site-packages\cx_Freeze\executable.py", line 66, in __init__
    self.base = base
    ^^^^^^^^^
  File "F:\dev\ayon-launcher\shim\.venv\Lib\site-packages\cx_Freeze\executable.py", line 114, in base
    raise OptionError(msg)
distutils.errors.DistutilsOptionError: no base named 'legacy/win32gui' ('legacy/win32gui-cp313-win_amd64.exe')
------------------------------------------
!!! Build failed. Check the log: .\shim\build.log
>>> Restoring current directory
```

---

The command `.\tools\manage.ps1 make-installer` fails due to be invalid/missing?

```
powershell .\tools\manage.ps1 make-installer

                    â–„â–ˆâ–ˆâ–„
         â–„â–ˆâ–ˆâ–ˆâ–„ â–€â–ˆâ–ˆâ–„ â–€â–ˆâ–ˆâ–€ â–„â–ˆâ–ˆâ–€ â–„â–ˆâ–ˆâ–€â–€â–€â–ˆâ–ˆâ–„    â–€â–ˆâ–ˆâ–ˆâ–„      â–ˆâ–„
        â–„â–„ â–€â–ˆâ–ˆâ–„  â–€â–ˆâ–ˆâ–„  â–„â–ˆâ–ˆâ–€ â–ˆâ–ˆâ–€      â–€â–ˆâ–ˆâ–„  â–„  â–€â–ˆâ–ˆâ–„    â–ˆâ–ˆâ–ˆ
       â–„â–ˆâ–ˆâ–€  â–ˆâ–ˆâ–„   â–€ â–„â–„ â–€  â–ˆâ–ˆ         â–„â–ˆâ–ˆ  â–ˆâ–ˆâ–ˆ  â–€â–ˆâ–ˆâ–„  â–ˆâ–ˆâ–ˆ
      â–„â–ˆâ–ˆâ–€    â–€â–ˆâ–ˆâ–„   â–ˆâ–ˆ    â–€â–ˆâ–ˆâ–„      â–„â–ˆâ–ˆâ–€  â–ˆâ–ˆâ–ˆ    â–€â–ˆâ–ˆ â–€â–ˆâ–€
     â–„â–ˆâ–ˆâ–€      â–€â–ˆâ–ˆâ–„  â–€â–ˆ      â–€â–ˆâ–ˆâ–„â–„â–„â–„â–ˆâ–ˆâ–€    â–ˆâ–€      â–€â–ˆâ–ˆâ–„

     Â·  Â· - =[ by YNPUT ]:[ http://ayon.ynput.io ]= - Â·  Â·

>>> Detecting host Python ... Resolved 119 packages in 3ms
Audited 49 packages in 1ms
OK [ 3.9 ]
Invoke-AyonInstaller : The term 'Invoke-AyonInstaller' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify
that the path is correct and try again.
At F:\dev\ayon-launcher\tools\manage.ps1:483 char:9
+         Invoke-AyonInstaller
+         ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Invoke-AyonInstaller:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

antirotor commented:
> The `build` failed for me on Windows, see:

I belive this is because you had already existing `.venv` under shim folder. Notice that it is trying to use pythpn 3.13 suddenly. When testing this, you need to clear all existing venvs just to be safe. We could add some handling logic, but it could be somewhat complex.

> Can we avoid having it force uv initialization in that case?

If you run `.\manage.ps1` without arguments, nothing about uv should be touched. Or at least it is not in my case?

> The `create-env` argument to `manage.ps1` still mentions Poetry

Right, removed that. I thought I did already.

> The command `.\tools\manage.ps1 make-installer` fails due to be invalid/missing

Yep, wrong function name, fixed too.

antirotor commented:
> But from within the virtual env - then hmm, no it now does not seem to do anything with `uv`, so that's good. Would be nice if that wouldn't need a Python to start.

~~This part is used to check for Python version. It is assumed you have python already installed and this just checks for the correct version. Arguably, this could be removed because `uv` is checking for that itself (or it can even manage the version itself). I'll remove the check.~~

It is actually used to check for AYON launcher version, just for display purposes. Could be probably parsed without the use of Python.

> ValueError: Can't find ISCC.exe

AYON still needs [Inno Setup](https://jrsoftware.org/isinfo.php) installed and in the `PATH`. We could check for it sooner (and provide more sensible error message)

antirotor commented:
> tbh, I'm not sure why it was looking inside `dependency_packages` in the first place. doesn't the uv thing prepare the environment before starting?

I don't believe this comes from this PR.  It fails in **ayon-core** - long after everything is processed by the launcher. Seems like `colorama` is used by `click` in ayon-core runtime dependency and is missing for some reason. `Click` is defined as a runtime depencecy of ayon-core and so it seems like fault of dependency package tool/poetry?

iLLiCiTiT commented:
> but where it was coming from? I cannot find it in the original pyproject.toml. This might come from different resolving strategies of uv/poetry maybe.

I guess from something like sphinx or mypy... poetry.lock did have it (you can check in develop)

MustafaJafar commented:
> but where it was coming from? I cannot find it in the original `pyproject.toml`. This might come from different resolving strategies of uv/poetry maybe.

I checked the log of dependencies tool, it mentioned installing colorama then
```
- colorama
Found existing installation: colorama 0.4.6
Uninstalling colorama-0.4.6:
  Successfully uninstalled colorama-0.4.6
```

Well, I do have this module installed for the same python version used inside AYON.
I can import it normally in any py script but not in AYON.

iLLiCiTiT commented:
> I checked the log of dependencies tool, it mentioned installing colorama then

You have to upload the AYON launcher build to your AYON server, and use that version of AYON launcher in bundle for which you're creating the dependency package to avoid uninstall.

iLLiCiTiT commented:
> For reference https://github.com/ynput/ayon-launcher/pull/280

You could just push it to this branch, or add comment here.
